### ISSUE TITLE
SC2154/check-unassigned-uppercase does not report error for $1,...

### ISSUE THREAD

**Issue by kkmuffme** [koalaman/shellcheck#3080]
#### For bugs
- Rule Id: SC2154/check-unassigned-uppercase
- My shellcheck version: 0.10.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
if [[ -n "$4" ]]
then
    echo "$5"
fi
```

#### Here's what shellcheck currently says:
No error

#### Here's what I wanted or expected to see:
>5 is referenced but not assigned

The global variables $1, $2,... (not $0) should be checked like any other uppercase variable when `check-unassigned-uppercase` is used.
The numbers are easily fatfingered and this is an easy mistake to make when restructuring code too.

**Comment by wileyhy** [comment]
Usually, positional parameters at the global scope are neither declared
nor assigned. They may be assigned with:

~ $ set -- x yz
  If a script writer plans for $4 to only ever be '--dir', and for $5 to be
defined if and only if $4 exists, then that would be some valid code there,
no?

Wiley

**Comment by wileyhy** [comment]
Well, maybe shellcheck could do something about that, I don't know.
Haskell's a undiscovered country if you ask me.

  I always assumed shellcheck wasn't able to check for assignment of global
positional parameters (which are assigned on the command line) because, as
a _static_ linter, there isn't any way for it to know the content of
_runtime_ data.  ?

  I can offer some workarounds, though. You might already be aware of these.

  A. "${1:-}" for silencing nounset.  (The ':=' parameter expansion cannot
be used for assigning values to positional parameters.)
  B. As a part of a battery of use tests:
    $ bash -u ./script.sh a b c
  C. Using `set -u` as part of some debugging procedures.
  D. "${2:?some error msg}" to check for null or unset variables.

  Wiley
